topic_message_query.go

```diff
@@ -115,77 +108,81 @@ func (query *TopicMessageQuery) Subscribe(client *Client, onNext func(TopicMessa
 		var err error
 
 		for {
-			if query.attempt <= query.maxAttempts && subClient == nil {
-				if query.limit != nil {
-					query.pb.Limit = *query.limit - query.counter
+			if err != nil {
+				if grpcErr, ok := status.FromError(err); ok {
+					if query.attempt < query.maxAttempts && query.retryHandler(err) {
+						subClient = nil
+
+						delay := math.Min(250.0*math.Pow(2.0, float64(query.attempt)), 8000)
+						time.Sleep(time.Duration(delay) * time.Millisecond)
+						query.attempt += 1
+					} else {
+						query.errorHandler(*grpcErr)
+						break
+					}
+				} else if err == io.EOF {
+					query.completionHandler()
+					break
+				} else {
+					panic(err)
 				}
+			}
 
+			if subClient == nil {
 				ctx, cancel := context.WithCancel(context.TODO())
```

Comment:
`cancel` should be called when a subscription finishes in the goroutine, regardless of success/failure